### ISSUE TITLE
 BinlogConnectorReplicator - replicator stopped at position

### ISSUE THREAD

**Issue by ypereirareis** [zendesk/maxwell#1491]
Hello !!!

(I hope everybody is fine.)

**Maxwell version:** v1.24.1
**Running with docker:** YES

We sometimes have the following error without particular (or identified) reasons.
The main problem is that our container remains **UP but Maxwell is no more doing anything**.

Do you have an idea ?

```
01:27:51,291 WARN  BinlogConnectorReplicator - replicator stopped at position: mysql-bin.091012:7257 -- restarting
Exception in thread "blc-192.168.xxx.xxx:3306" java.lang.IllegalStateException: BinaryLogClient is already connected
	at com.github.shyiko.mysql.binlog.BinaryLogClient.connect(BinaryLogClient.java:494)
	at com.github.shyiko.mysql.binlog.BinaryLogClient$7.run(BinaryLogClient.java:857)
	at java.lang.Thread.run(Thread.java:748)
```

Thanks to all core devs and contributors.

**Comment by lucastex** [comment]
Do you have more than one Maxwell connected on this same database?

**Comment by ypereirareis** [comment]
* I have two completely distinct Maxwell instances, connected on two different databases, on the same MySQL server.
* But sometimes, my docker policy restart the container because of CPU or RAM usage. It seems that when restarting, the previous connection is still there or something like that. If I restart the container again after a few seconds, everything is OK.

**Comment by ypereirareis** [comment]
* I also have other SQL slaves connected to this database for replication. But I use different server ids.

**Comment by lucastex** [comment]
Can you check if replica server ids is defined with different values for each Maxwell server?

**Comment by osheroff** [comment]
@lucastex It's somewhat interesting *why* the binlog replicator is disconnecting, but more interesting is looking deeper into what's really crashing maxwell.  Maxwell assumes connections will break, but if you chase this guy down the rabbit hole you'll find this:

https://github.com/zendesk/maxwell/blob/564a32f24b71b46ef3d84b271557a68dc757ce3f/src/main/java/com/zendesk/maxwell/replication/BinlogConnectorReplicator.java#L408-L428

on line 428 we try to restart a dead replicator connection, but oddly the replicator throws an error, which, when chased down, appears to be a race condition: 

https://github.com/osheroff/mysql-binlog-connector-java/blob/1a00ac240a59d6504744c3f75e5f3f0256897be4/src/main/java/com/github/shyiko/mysql/binlog/BinaryLogClient.java#L499-L502

(the only place `connectLock` is used is inside the `connect` function itself).  

So what seems to be happening is:

- the user *probably* has a fairly flaky setup, where the replicator's connection gets dropped often.  I don't know why this happens.  servers sometimes get feisty.
- maxwell notices that the connection is dead before binlog-connector has exited the `connect` function, and tries to reconnect.
- everything goes fucked.

We should fix it on either maxwell's side or binlog-connector's side; in an ideal world I'd rather fix it binlog-connector side, so it doesn't tell the client code that it's disconnected until it's really ready to reconnect.  But it's a hell of a lot simpler on maxwell's side, the paths are just less tangled up.  on maxwell's side the fix would be to not rely on the `isConnected` function but instead listen for a disconnection event via the "lifecycle listener" stuff; we already have it mostly wired up.

**Comment by ypereirareis** [comment]
@lucastex Yes all my server ids are different.
@osheroff Hope you'll succeed in fixing the problem. :)

Thx

**Comment by osheroff** [comment]
fixes for the race condition went out in 1.26.3

**Comment by ypereirareis** [comment]
Thx, gonna try it ASAP.
